### DISCUSSION TITLE
[SSR] Declarative Shadow DOM and Cross Browser Support ("static" SSR)

### DISCUSSION THREAD

**Discussion by thescientist13** [lit/lit#2589] [Q&A] [answered]
## Overview
Maybe this is ultimately my misunderstand of Declarative Shadow DOM and / or browser handling of unknown tags, but the scenario here is that I have been playing around with LIt SSR in the context of prerendering Lit components as static HTML, trying to ship no JavaScript if possible.

I am asking here since Lit+SSR outputs using Declarative Shadow DOM and so was hoping the community could help me out.  🙏 

## Example
For example, given this output of HTML generated by server-rendering with the LIt SSR package
```html
<!--lit-part BRUAAAUVAAA=--><!--lit-part GqZOtE8VqOY=--><!DOCTYPE html>
<html lang="en" prefix="og:http://ogp.me/ns#">
   <head>
      <title>Tuesday's Tunes</title>
   </head>
   <body>
      <section>
         <tt-header>
            <template shadowroot="open">
               <style>
                  h1 {
                  width: 90%;
                  margin: 30px auto;
                  text-align: center;
                  }
               </style>
               <!--lit-part 1dQVzEiXVpM=-->
               <header>
                  <h1>Coming Soon!</h1>
               </header>
               <!--/lit-part-->
            </template>
         </tt-header>
      </section>
      <section>
         <main>
            <tt-banner>
               <template shadowroot="open">
                  <style>
                     img {
                     display: block;
                     width: 50%;
                     margin: 0 auto 20px;
                     }
                  </style>
                  <!--lit-part bzQrwU2cJWo=-->
                  <div>
                     <img src="/assets/tuesdays-tunes.jpg"/>
                  </div>
                  <!--/lit-part-->
               </template>
            </tt-banner>
         </main>
      </section>
      <section>
         <tt-footer>
            <template shadowroot="open">
               <style>
                  footer {
                  width: 90%;
                  margin: 0 auto;
                  text-align: center;
                  color: white;
                  }
               </style>
               <!--lit-part CwV44/uBt1A=-->
               <footer>
                  <slot></slot>
                  <h5>
                     &copy; <!--lit-part-->2022<!--/lit-part--> - Tuesday's Tunes / Analog Studios
                  </h5>
               </footer>
               <!--/lit-part-->
            </template>
            <a href="https://www.youtube.com/user/analogstudios">YouTube</a> |
            <a href="https://www.facebook.com/dave.flamand.1426">Facebook</a> |
            <a href="https://open.spotify.com/playlist/1jOFrovuxCGulkGdUhR5f7?si=R1ZowQxqSRefj6HLOasSEQ&nd=1">Spotify</a>
         </tt-footer>
      </section>
   </body>
</html>
<!--/lit-part--><?><!--/lit-part-->
```

In Chrome / Edge, all the expected content is visible
<img width="1089" alt="Screen Shot 2022-02-28 at 21 21 01" src="https://user-images.githubusercontent.com/895923/156092486-27af8de0-59f2-4700-a141-c98c1bc1f48b.png">

But in Firefox / Safari, only the `<slot>`ed content is visible 
<img width="473" alt="Screen Shot 2022-02-28 at 21 20 50" src="https://user-images.githubusercontent.com/895923/156092505-a1c1a621-3f2b-4b1b-9bfc-db1a8df22ede.png">

----

If I add [the polypill](https://web.dev/declarative-shadow-dom/#polyfill), then the content appears as expected in Firefox / Safari
```js
<script>
  if (!HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
    (function attachShadowRoots(root) {
      root.querySelectorAll("template[shadowroot]").forEach(template => {
        const mode = template.getAttribute("shadowroot");
        const shadowRoot = template.parentNode.attachShadow({ mode });
        shadowRoot.appendChild(template.content);
        template.remove();
        attachShadowRoots(shadowRoot);
      });
    })(document);
  }
</script>
```

## Questions
I understand that DSD is only supported natively by Chromium based browsers, but it was also my understanding that browsers would generally ignore tags they didn't understand and render what they do?

So even for a block of HTML like this
```html
<tt-footer>
  <template shadowroot="open">
     <style>
        footer {
        width: 90%;
        margin: 0 auto;
        text-align: center;
        color: white;
        }
     </style>
     <!--lit-part CwV44/uBt1A=-->
     <footer>
        <slot></slot>
        <h5>
           &copy; <!--lit-part-->2022<!--/lit-part--> - Tuesday's Tunes / Analog Studios
        </h5>
     </footer>
     <!--/lit-part-->
  </template>
  <a href="https://www.youtube.com/user/analogstudios">YouTube</a> |
  <a href="https://www.facebook.com/dave.flamand.1426">Facebook</a> |
  <a href="https://open.spotify.com/playlist/1jOFrovuxCGulkGdUhR5f7?si=R1ZowQxqSRefj6HLOasSEQ&nd=1">Spotify</a>
</tt-footer>
```

I would still at least expect the static content in the `<h5>` to show up?

1. Do browser's not actually work this way?  Or is it because the `<h5>` is nested inside a custom element `<tt-footer>` the browser doesn't know, and so then the content is "swallowed" up in that?
1. Or is it that these browser's do understand `<template>` but _not_ `shadowroot="open"` and so that is where the content is getting lost?

It's great the polyfill works but curious if there is a way to get this content out to a browser without any JavaScript at all to get the content to appear?

**Comment by justinfagnani** [comment, marked as answer]
`<template>` is implemented natively on Safari and Firefox, and the behavior of `<template>` before support for the `shadowroot` attribute is to move the contents into a document fragment, which will not be rendered.

So the behavior in Safari/Firefox is exactly as if you had removed the `shadowroot` attribute from all the `<template>`s.

**Reply by thescientist13** [reply]
Thanks @justinfagnani and indeed, first line on MDN 🤦 😅 
> The `<template>` [HTML](https://developer.mozilla.org/en-US/docs/Web/HTML) element is a mechanism for holding [HTML](https://developer.mozilla.org/en-US/docs/Glossary/HTML) that is not to be rendered immediately when a page is loaded but may be instantiated subsequently during runtime using JavaScript.

So basically until these browsers actually support DSD, this is the behavior to expect it seems.

----

On a related note, I just now tried implementing `createRenderRoot` as a test to see if that could help break "out" the HTML but it still seems like the same output as before (with no more HTML visible than before)?
```js
import { css, html, LitElement } from 'lit';

class FooterComponent extends LitElement {

  static styles = css`
    footer {
      width: 90%;
      margin: 0 auto;
      text-align: center;
      color: white;
    }
  `;

  createRenderRoot() {
    return this;
  }

  render() {
    const year = new Date().getFullYear();

    return html`
      <footer>
        <slot></slot>
        <h5>&copy; ${year} - Tuesday's Tunes / Analog Studios</h5>
      </footer>
    `;
  }
}

customElements.define('tt-footer', FooterComponent);
```

```html
<tt-footer>
  <template shadowroot="open">
    <style>
      footer {
      width: 90%;
      margin: 0 auto;
      text-align: center;
      color: white;
      }
    </style>
    <!--lit-part CwV44/uBt1A=-->
    <footer>
      <slot></slot>
      <h5>
        &copy; <!--lit-part-->2022<!--/lit-part--> - Tuesday's Tunes / Analog Studios
      </h5>
    </footer>
    <!--/lit-part-->
  </template>
  <a href="https://www.youtube.com/user/analogstudios">YouTube</a> |
  <a href="https://www.facebook.com/dave.flamand.1426">Facebook</a> |
  <a href="https://open.spotify.com/playlist/1jOFrovuxCGulkGdUhR5f7?si=R1ZowQxqSRefj6HLOasSEQ&nd=1">Spotify</a>
</tt-footer>
```

Not sure if I am doing something wrong there in the implementation but would that same output be expected?  I know `<slot>` only works with DSD but just curious what sort of escape hatches might be available in the interim while these other browsers catch up in their support for DSD to drive this static HTML prerendering like workflow.

Thanks!
